### ISSUE TITLE
Inconsistent conditional type results between interfaces and inline types when matching against an index type signature

### ISSUE THREAD

**Issue by greyepoxy** [Microsoft/TypeScript#24381]
**TypeScript Version:**  3.0.0-dev.20180522

**Search Terms:**  Index, Conditional, Signature

**Code**

```ts
type IndexSignatureType<T> = T extends { [key: string]: infer R } ? R : never;

interface TestInterface {propWithString: string; propWithNumber: number }

type testWithInterface = IndexSignatureType<TestInterface>;
// type testWithInterface = never

type testWithInlineType = IndexSignatureType<{ propWithString: string; propWithNumber: number }>;
// type testWithInlineType = string | number
```

**Expected behavior:**
Both `testWithInterface` and `testWithInlineType` should report that they are of the `never` type

**Actual behavior:**
`testWithInterface` reports it is of the `never` type but `testWithInlineType` reports that it is a union of its possible props

**Playground Link:**  
[link](http://www.typescriptlang.org/play/#src=type%20IndexSignatureType%3CT%3E%20%3D%20T%20extends%20%7B%20%5Bkey%3A%20string%5D%3A%20infer%20R%20%7D%20%3F%20R%20%3A%20never%3B%0D%0A%0D%0Ainterface%20TestInterface%20%7BpropWithString%3A%20string%3B%20propWithNumber%3A%20number%20%7D%0D%0A%0D%0Atype%20testWithInterface%20%3D%20IndexSignatureType%3CTestInterface%3E%3B%0D%0A%2F%2F%20type%20testWithInterface%3D%20never%0D%0A%0D%0Atype%20testWithInlineType%20%3D%20IndexSignatureType%3C%7B%20propWithString%3A%20string%3B%20propWithNumber%3A%20number%20%7D%3E%3B%0D%0A%2F%2F%20type%20testWithInlineType%20%3D%20string%20%7C%20number)

**Related Issues:** 
There seemed to be several issues related to index signatures. Not sure if they are really related or not... for example https://github.com/Microsoft/TypeScript/issues/23226 and https://github.com/Microsoft/TypeScript/issues/23994

**Context:**
I am working on an `any` object validation library, where a user provides an interface as a generic to a function and then has to provide an object with "validation functions" for each property in the given interface. The code then uses this property to validation function object to validate that an arbitrary object looks like the given interface. Works great until index signature types get involved. That lead me to using the described `IndexSignatureType` conditional generic type above to figure out if a given interface/type had an index signature type as part of its definition or not. My plan was to use this type definition as a mechanism for forcing the user to use a different function in this case where they just provide a single validation function for all of the object props. With the above inconsistencies in the conditional types this does not work.
Some examples of what I am talking about,
```ts
import * as validate from 'ts-any-validator'; // my library, not published though sorry...

interface TestInterface {
  propWithString: string;
  propWithNumber: number;
}

const validateIsTestInterface = validate.isObjectWithKnownProps<TestInterface>({
  propWithString: validate.isString,
  propWithNumber: validate.isNumber,
});
// This this an example usage which throws an error if validation for any of the properties fails
validateIsTestInterface({ propWithString: '', propWithNumber: 10 });

// Using a mechanism similar to the one described above neither of these works, which is great!
// In both cases below there is an error because the param resolves to a never (exactly what I want)
interface AnotherTestInterface {
  [key: string]: string;
}
const validateIsAnotherTestInterface = validate.isObjectWithKnownProps<AnotherTestInterface>({}); // expected error Argument is not assignable to 'never'

const validateIsLikeAnotherTestInterface = validate.isObjectWithKnownProps<{
  [key: string]: string;
}>({}); // expected error Argument is not assignable to 'never'

// Instead of using `isObjectWithKnownProps` for types with signature types a different function is used
// no error in this case
const validateIndexSignatureType = validate.isObjectWithIndexSignature<AnotherTestInterface>(validate.isString);


// But this one should not resolve to a never and does because of this issue described above
const validateHasProp = validate.isObjectWithKnownProps<{
  prop: string;
}>({}); // unexpected error Argument is not assignable to 'never'
```
Hopefully the context makes some sense...

Thank you for all your hard work!

FYI @mjbvz

**Comment by jack-williams** [comment]
Is this a small reproduction of the problem? (sans conditional types)

```ts
interface TestInterface {propWithString: string; propWithNumber: number }
type Check<T extends { [key: string]: {} }> = T;

type A = Check<{ propWithString: string; propWithNumber: number }>; // Ok
type B = Check<TestInterface>; // Error
```

**Comment by mhegazy** [comment]
Duplicate of https://github.com/Microsoft/TypeScript/issues/15300

**Comment by greyepoxy** [comment]
@mhegazy 
Yeah this does seem related to that but looks like I am asking for something different then folks in that thread.

From what I can tell folks in that issue are asking for "Things that look like `{ [key: string]: string; }` should be assignable/castable to that type." Today you can do that with types but not with interfaces.
That seems to explain the inconsistency I am calling out here.


I guess what I am really asking for is "How can I conditionally match on a type/interface that has an explicit type index signature." My guess is that if https://github.com/Microsoft/TypeScript/issues/15300 is resolved its going to mean in my above example both `testWithInterface` and `testWithInlineType` resolve to `string | number` which is the opposite of what I was hoping for.

hmm is this a feature request then?

**Comment by mhegazy** [comment]
Type literals have an implicit index signatures. interfaces do not. that is the current design. see https://github.com/Microsoft/TypeScript/issues/15300#issuecomment-332366024 for more details.  #15300 tracks addressing the discrepancy.. do not know which way it would go.

There is no way to distinguish between an implicit index signature, and an implicit one. and again this is by design.

**Comment by greyepoxy** [comment]
okay closing this out then, might or might not open a feature request for conditionally matching support against an explicit index signature.
